gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspSubplugin.kt

```diff
@@ -75,29 +82,116 @@ class KspGradleSubplugin @Inject internal constructor(private val registry: Tool
+            }
+        }
     override fun apply(project: Project) {
         project.extensions.create("ksp", KspExtension::class.java)
-        project.configurations.create(KSP_CONFIGURATION_NAME)
-
+        artifactName = if (project.hasProperty("KSP_ARTIFACT_NAME")) {
+            project.properties.get("KSP_ARTIFACT_NAME") as String
+        } else {
+            DEFAULT_KSP_ARTIFACT_NAME
+        }
+        project.pluginManager.withPlugin("org.jetbrains.kotlin.jvm") {
+            // kotlin extension has the compilation target that we need to look for to create configurations
+            decorateKotlinExtension(project)
+        }
+        project.pluginManager.withPlugin("com.android.application") {
+            // for android apps, we need a configuration per source set
+            decorateAndroidExtension(project)
+        }
+        project.pluginManager.withPlugin("com.android.library") {
+            // for android libraries, we need a configuration per source set
+            decorateAndroidExtension(project)
+        }
         registry.register(KspModelBuilder())
     }
 
+    private fun decorateKotlinExtension(project:Project) {
+        project.extensions.configure(KotlinSingleTargetExtension::class.java) { kotlinExtension ->
+            kotlinExtension.target.compilations.createKspConfigurations(project) { kotlinCompilation ->
+                kotlinCompilation.kotlinSourceSets.map {
+                    it.kspConfigurationName
+                }
+            }
+        }
+    }
+
+    private fun decorateAndroidExtension(project:Project) {
+        @Suppress("UnstableApiUsage")
+        project.extensions.configure(CommonExtension::class.java) {
+            it.sourceSets.createKspConfigurations(project) { androidSourceSet ->
+                listOf(androidSourceSet.kspConfigurationName)
+            }
+        }
+    }
+
+    /**
+     * Creates a KSP configuration for each element in the object container.
+     */
+    private fun<T> NamedDomainObjectContainer<T>.createKspConfigurations(
+        project: Project,
+        getKspConfigurationNames : (T)-> List<String>
+    ) {
+        val mainConfiguration = project.configurations.maybeCreate(KSP_MAIN_CONFIGURATION_NAME)
+        all {
+            getKspConfigurationNames(it).forEach { kspConfigurationName ->
+                if (kspConfigurationName != KSP_MAIN_CONFIGURATION_NAME) {
+                    val existing = project.configurations.findByName(kspConfigurationName)
+                    if (existing == null) {
+                        project.configurations.create(kspConfigurationName) {
+                            it.extendsFrom(mainConfiguration)
+                        }
+                    }
+                }
+            }
+        }
+    }
+
     override fun isApplicable(kotlinCompilation: KotlinCompilation<*>): Boolean = true
 
     override fun applyToCompilation(kotlinCompilation: KotlinCompilation<*>): Provider<List<SubpluginOption>> {
```

Comment:
You also need to register generated:
- java sources
- class files
with AGP. See https://github.com/JetBrains/kotlin/blob/41c4693ebf4f394b58a66164ddc206ad82ce1096/libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/kapt/Kapt3KotlinGradleSubplugin.kt#L475.

Kotlin sources do not need to registered asn they are anyhow picked up by kotlin compile task.